### ISSUE TITLE
Make @Compressed work together with @CacheResult

### ISSUE THREAD

**Issue by rafaelfnx** [quarkusio/quarkus#42975]
### Description

At the request of **Georgios Andrianakis** in [this topic](https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/Does.20.40Compressed.20with.20.40CacheResult.20work.20together.3F), I'm opening this issue so we can monitor updates to the following question _**"Does @Compressed with @CacheResult work together?"**_.

And if we have the possibility of implementing this, if it is possible and practical.

An example of code where this improvement should take effect:
```java
    @GET
    @Path("/{fileName}")
    @Compressed
    @CacheResult(cacheName = "image-cache")
    public Uni<Buffer> getImage(@PathParam("fileName") String fileName) {
        return imageService.getImage(fileName);
    }
```

Should @CacheResult store the already compressed result since the method was also annotated with @Compressed?

**Martin Kouba** said in our topic, that this doesn't happen, as @CacheResult is implemented as a method interceptor, while @Compressed is a hint for the underlying http framework to add a header informing compression (whereas Vertx actually compresses).

Since both annotations are used by very distant components, would it be possible and practical to make them interact?

**Comment by geoand** [comment]
cc @mkouba

**Comment by mkouba** [comment]
I don't think it's feasible unless we rework the current implementation completely but maybe I'm missing something.

In theory, we could introduce a special interceptor that would compress the result before the `CacheResultInterceptor` puts the value in the cache but it could be even more confusing. First of all, I don't think that Vert.x/Netty exposes the compression API; in other words, we would need to use a different API (and bypass the Vert.x compression config). Also if the cached value is already compressed then Vert.x would attempt to compress the compressed data again because we can't disable the compression so that the correct HTTP headers are set etc.

CC @cescoffier

**Comment by rafaelfnx** [comment]
I think that following your idea, it would then be necessary NOT to activate vertx compression as is done in the [documentation](https://vertx.io/docs/4.4.9/vertx-core/java/#:~:text=To%20enable%20compression,is%20not%20enabled), and do the compression in another way, before the `CacheResultInterceptor`, right? 

It might still be possible to create a specific caching mechanism for HTTP responses instead of using a method interceptor, which would store the already compressed http response in the cache, maybe like a JAX-RS write interceptor, but this would also involve compressing in another way and disable vertx compression so that it would not compress again.

>  First of all, I don't think that Vert.x/Netty exposes the compression API;

Maybe it would even make sense to open a discussion in the vertx repository considering the exposure of the compression API, right? This way, it would be possible to keep vertx compression active and avoid doing it any other way.

In my opinion, I think that if Vertx really decided to expose the compression API, the second way I suggested, of creating a compression mechanism at the JAX-RS level would be interesting.

But I just gave ideas here, as it seems that any path to make this possible takes a lot of effort, planning and a major modification of the current structure, so for now I'll be content with just keeping the content not yet compressed in the cache.

Thanks @mkouba for the general explanation!

Do you think this issue can now be closed? Or kept open to discussing this idea again in the future?

**Comment by mkouba** [comment]
> I think that following your idea, it would then be necessary NOT to activate Vertx compression as is done in the [documentation](https://vertx.io/docs/4.4.9/vertx-core/java/#:~:text=To%20enable%20compression,is%20not%20enabled), and do the compression in another way, before the `CacheResultInterceptor`, right?

Yes, you would need a special CDI interceptor that would be called after the `CacheResultInterceptor` (but before the intercepted method) and perform the compression in this interceptor. This way the cached value would be compressed. But you would still need to modify the HTTP response, i.e. add the `Content-Encoding: gzip` header manually. I'm not 100% sure how this would work though.

FYI the [GZip support](https://quarkus.io/guides/resteasy#gzip-support) for RESTEasy Classic in Quarkus kind of works  like that. AFAIK they use a custom `jakarta.ws.rs.ext.WriterInterceptor`/`jakarta.ws.rs.ext.ReaderInterceptor`.

> In my opinion, I think that if Vertx really decided to expose the compression API, the second way I suggested, of creating a compression mechanism at the JAX-RS level would be interesting. This approach would not only handle compression but also store the results already transformed into JSON, rather than simply caching the object returned by the method and repeatedly converting it to JSON.

I think that Vert.x is reusing some Netty stuff for compression. Not sure.

> But I just gave ideas here, as it seems that any path to make this possible takes a lot of effort, planning and a major modification of the current structure, so for now I'll be content with just keeping the content not yet compressed in the cache.
> 
> Thanks @mkouba for the general explanation!
> 
> Do you think this issue can now be closed? Or kept open to discussing this idea again in the future?

Let's keep it open for now.

**Comment by rafaelfnx** [comment]
> FYI the [GZip support](https://quarkus.io/guides/resteasy#gzip-support) for RESTEasy Classic in Quarkus kind of works like that. AFAIK they use a custom jakarta.ws.rs.ext.WriterInterceptor/jakarta.ws.rs.ext.ReaderInterceptor.

Wow, very interesting to know!

> Let's keep it open for now.

It was more useful than I imagined, because now I had a small related question, due to my need to customize the response headers, I needed to change the return of my method, now it is as follows:

```java
    @GET
    @Path("/{fileName}")
    @Compressed
    @CacheResult(cacheName = "image-cache")
    public Uni<Response> getImage(@PathParam("fileName") String fileName) {
        return imageService.getImage(fileName);
    }
```

And now, the `Buffer` that I returned before is contained within the `Response`.

But my question is, will this work? `Response` implements the `AutoCloseable` interface, I don't know if this object can be reused from the cache or I always need to create a new instance that is not closed.

Thank you in advance!

**Comment by mkouba** [comment]
> But my question is, will this work? `Response` implements the `AutoCloseable` interface, I don't know if this object can be reused from the cache or I always need to create a new instance that is not closed.

I have no idea. CC @geoand

**Comment by geoand** [comment]
I believe it should work fine, but I have not tried it to be honest
